specification/src/main/asciidoc/webprofile/jpa-cdi/CDI-JPA.adoc

```diff
@@ -0,0 +1,120 @@
+:cdi-spec: https://jakarta.ee/specifications/cdi/4.1
+:jpa-spec: https://jakarta.ee/specifications/persistence/3.1/jakarta-persistence-spec-3.1
+
+=== Jakarta Persistence & Jakarta Context and Dependency Injection (CDI) Integration
+This section specifies the requirements for Jakarta EE environments that integrate both {jpa-spec}[Persistence] and {cdi-spec}[CDI].
+
+==== Obtaining an Entity Manager using CDI
+
+A container environment might feature built-in integration of
+Jakarta Persistence with the CDI bean manager, allowing injection
+of a container-managed entity manager using the annotation
+_jakarta.inject.Inject_. In the Jakarta EE environment, the
+container must feature built-integration of Jakarta Persistence
+with the CDI bean manager.
+
+For each persistence unit, a container which features such
+built-in integration must make available a bean with:
+
+- bean type _EntityManager_,
+- the qualifiers specified by _qualifier_ XML elements in
+_persistence.xml_, or _jakarta.enterprise.inject.Default_,
+if no qualifiers are explicitly specified,
+- the scope specified by the _scope_ XML element in
+_persistence.xml_, or _jakarta.transaction.TransactionScoped_
+if no scope is explicitly specified,
+- no interceptor bindings,
```

Comment:
What is the intend here? That just by default there are no interceptor bindings, or, that it is forbidden for interceptors to be defined for this type?

Reply:
If we allow to supply a qualifier to avoid some migration concerns, we need an example on how to specify the qualifier in the `persistence.xml`. I guess the .xsd needs to be updated. Is this too late since Jakarta Persistence has been released. I think we should just say the `Default` qualifier will be used. I am not sure how you can add scope either in the `persistence.xml` unless it was done already.

Reply:
See the later section on the explanation. Please add a reference to the later section on the xsd.

Reply:
> That just by default there are no interceptor bindings, or, that it is forbidden for interceptors to be defined for this type?


@arjantijms I'm not sure I understand what you mean by this question, because interceptor bindings belong to beans, not to bean _types_.

 Of course one could define a bean with type `Entity Manager ` and interceptor bindings, but the bean defined _here_ does not have any interceptor bindings, because we don't provide any place to specify such bindings. (We would need to add an element to `persistence.xml`.)

Reply:
> > That just by default there are no interceptor bindings, or, that it is forbidden for interceptors to be defined for this type?
> 
> @arjantijms I'm not sure I understand what you mean by this question, because interceptor bindings belong to beans, not to bean _types_.


Yes, though the text now says:

```
For each persistence unit, a container which features such built-in integration must make available a *bean* with:

no interceptor bindings,
```

I would expect the spec to say what IS required. For things that are NOT required, I would expect those things to be either left out or explicitly saying they are not required.

The way it's written now would probably trigger a TCK engineer to write a test that very specifically checks in whatever way is possible that the injected EntityManager does absolutey not have any interceptor bindings.

If my implementation would now for some reason add interceptor bindings for internal use, say for tracing, or for transaction management or whatever, then this would not be allowed, seemingly. But then the question arises; WHY is this restriction in place?

But from your reply I taste you dont want to specificallty constraint the Bean to not have interceptor bindings, but just that by default it does not need any (spec defined) interceptor bindings.

>  because we don't provide any place to specify such bindings. 

Isn't that already available via an extension? I quite often programmatically add interceptor bindings to beans that way.

Reply:
> If my implementation would now for some reason add interceptor bindings for internal use, say for tracing, or for transaction management or whatever, then this would not be allowed, seemingly. But then the question arises; WHY is this restriction in place?

If the bean has interceptor bindings that are visible to the application developer, then that would result in non-portable behavior. Now, of course, an implementation could in principle have its own internal interceptor binding types, and the bean could in principle have such interceptor bindings. But then such binding types would not be visible to the application programmer, nor to a TCK engineer.

Reply:
> Isn't that already available via an extension? I quite often programmatically add interceptor bindings to beans that way.

I mean, that's fine, nothing in that language says that _portable extensions_ can't add interceptor bindings. Just the bean provided by the container does not have any.

Reply:
> Just the bean provided by the container does not have any.

I hear you, though as a container I can use a portable extensions just as well of course.

The question is perhaps how it would result in non-portable behaviour. Say I would in GlassFish add an interceptor binding to the `Bean<T>` for `EntityManager#persist`,  to track how many times persist is called (I display a nice graph of this in the admin console).

Applications injecting `EntityManager` on my GF instace will have a working `persist` method, and that same code would work on WildFly just as well. Tracking persists calls is just an internal GF feature.

Reply:
Well the application program or framework could associate an interceptor with the given binding type, and then that interceptor would be triggered in this container but not in other containers.